File: app/src/main/java/com/example/proif_a01_java/View/ProductDetailsFragments.java
package com.example.proif_a01_java.View;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Color;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.constraintlayout.helper.widget.Carousel;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentResultListener;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.example.proif_a01_java.Model.Page;
import com.example.proif_a01_java.Model.Product;
import com.example.proif_a01_java.R;
import com.example.proif_a01_java.databinding.FragmentProductDetailsBinding;

import org.parceler.Parcels;

import java.util.ArrayList;

public class ProductDetailsFragments extends Fragment implements View.OnClickListener{
    private FragmentProductDetailsBinding binding;
    private FragmentManager fragmentManager;

    // constructor kosong
    public ProductDetailsFragments () {

    }

    // singleton
    public static ProductDetailsFragments newInstance() {
        ProductDetailsFragments fragments = new ProductDetailsFragments();
        return fragments;
    }

    @Override
    public View onCreateView (LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        this.fragmentManager = getParentFragmentManager();
        this.binding = FragmentProductDetailsBinding.inflate(inflater, container, false);

        // from fragment list and tiles listener
        this.fragmentManager.setFragmentResultListener(
                "MOVE_DETAILS", this, new FragmentResultListener() {
                    @Override
                    public void onFragmentResult(@NonNull String requestKey, @NonNull Bundle result) {
                        Product product = (Product) Parcels.unwrap(result.getParcelable("products"));
                        setProducts(product);
                    }
                }
        );

        // set onclick listener
        this.binding.btnAdd.setOnClickListener(this::onClick);
        this.binding.ivBack.setOnClickListener(this::onClick);

        return this.binding.getRoot();
    }

    public void setProducts (Product product) {
        // set all products info here
        this.binding.tvName.setText(product.name);
        this.binding.tvCategory.setText(product.category);
        this.binding.tvCondition.setText(product.condition);
        this.binding.tvPrice.setText(product.price + "");
        this.binding.tvDesc.setText(product.description);

        // sementara belum carousel, !PERLU DIUBAH JADI BENTUK CAROUSEL!
        //IMPLEMENTASI GLIDE LIBRARY
//        Glide.with(getActivity())
//                .load(getActivity().getResources().getIdentifier(product.photo,"drawable", getActivity().getPackageName()))
//                .diskCacheStrategy(DiskCacheStrategy.NONE)
//                .placeholder(R.drawable.progress_bar)
//                .into(this.binding.ivPics);

        int numOfPhotos=product.photo.size();
        int image[]=new int[numOfPhotos];
        for(int i=0;i<numOfPhotos;i++){
            image[i]=getResources().getIdentifier(product.photo.get(i),"drawable", getActivity().getPackageName());
        }
        this.binding.ivPics0.setImageResource(image[0]);


        this.binding.carousel.setAdapter(new Carousel.Adapter() {
            @Override
            public int count() {
                return image.length;
            }

            @Override
            public void populate(View view, int index) {
                Bitmap bmp=BitmapFactory.decodeResource(getResources(),image[index]);
                BitmapDrawable bmp2= new BitmapDrawable(getResources(),bmp);
                view.setBackground(bmp2);


            }

            @Override
            public void onNewItem(int index) {

            }
        });


    }

    @Override
    public void onClick(View view) {
        if (view == this.binding.btnAdd) {
            this.addCartToast();
        }
        else if (view == this.binding.ivBack) {
            this.changePage(Page.PAGE_LIST_MODE);
        }
    }

    private void addCartToast () {
        Toast toast = Toast.makeText(getContext(), "Product added to cart", Toast.LENGTH_SHORT);
        toast.show();
    }

    //METHOD GANTI HALAMAN
    private void changePage(int page){
        Bundle result = new Bundle();
        result.putInt("page", page);
        this.fragmentManager.setFragmentResult("CHANGE_PAGE", result);
    }
}

File: app/src/main/java/com/example/proif_a01_java/View/ProductDetailsFragments.kt
package com.example.proif_a01_java.View

import android.view.View.OnTouchListener
import com.example.proif_a01_java.Model.Product
import android.view.LayoutInflater
import android.view.ViewGroup
import android.os.Bundle
import org.parceler.Parcels
import android.util.Log
import android.view.MotionEvent
import android.view.View
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import android.widget.Toast
import androidx.fragment.app.Fragment

import com.example.proif_a01_java.Model.Page

import com.example.proif_a01_java.databinding.FragmentProductDetailsBinding

class ProductDetailsFragments: Fragment(), View.OnClickListener, OnTouchListener {

    lateinit var binding: FragmentProductDetailsBinding
    private var x1 = 0f
    private var x2 = 0f
    private var curPic = 0
    lateinit var curProduct: Product
    private var pageFrom: Int = Page.PAGE_LIST_MODE

    companion object {
        fun newInstance(): ProductDetailsFragments {
            return ProductDetailsFragments()
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {

        binding = FragmentProductDetailsBinding.inflate(inflater, container, false)

        // from fragment list and tiles listener
        parentFragmentManager.setFragmentResultListener(
            "MOVE_DETAILS", this
        ) { requestKey, result ->
            pageFrom = result.getInt("pageFrom")
            val product = Parcels.unwrap<Any>(result.getParcelable("products")) as Product
            setProducts(product)
        }

        // set onclick listener
        binding.btnAdd.setOnClickListener { view: View -> onClick(view) }
        binding.ivBack.setOnClickListener { view: View -> onClick(view) }
        binding.ivPics.setOnTouchListener { view: View, motionEvent: MotionEvent ->onTouch(view,motionEvent)}
        return binding.root
    }

    fun setProducts(product: Product) {
        // set all products info here
        binding.tvName.text = product.name
        binding.tvCategory.text = product.category
        binding.tvCondition.text = product.condition
        binding.tvPrice.text = product.price.toString() + ""
        binding.tvDesc.text = product.description
        curProduct = product

        //IMPLEMENTASI GLIDE LIBRARY
        Glide.with(requireActivity())
            .load(
                requireActivity().resources.getIdentifier(
                    product.photo[0],
                    "drawable",
                    requireActivity().packageName
                )
            )
            .diskCacheStrategy(DiskCacheStrategy.NONE)
            .into(binding.ivPics)
    }

    override fun onClick(view: View) {
        if (view === binding.btnAdd) {
            addCartToast()
        } else if (view === binding.ivBack) {
            changePage(pageFrom)
        }
    }

    private fun addCartToast() {
        val toast = Toast.makeText(context, "Product added to cart", Toast.LENGTH_SHORT)
        toast.show()
    }

    //METHOD GANTI HALAMAN
    private fun changePage(page: Int) {
        val result = Bundle()
        result.putInt("page", page)
        parentFragmentManager.setFragmentResult("CHANGE_PAGE", result)
    }

    override fun onTouch(view: View, motionEvent: MotionEvent): Boolean {
        when (motionEvent.action) {
            MotionEvent.ACTION_DOWN -> {
                x1 = motionEvent.x
                Log.d("DOWN", x1.toString() + "")
            }
            MotionEvent.ACTION_UP -> {
                x2 = motionEvent.x
                Log.d("UP", x2.toString() + "")
                if (Math.abs(x2 - x1) > 150) {
                    if (x2 > x1) {
                        // swipe right
                        curPic++
                        if (curPic > 2) {
                            curPic = 0
                        }
                    } else {
                        // swipe left
                        curPic--
                        if (curPic < 0) {
                            curPic = 2
                        }
                    }
                    slidePic()
                }
            }
        }
        return true
    }

    private fun slidePic() {
        Glide.with(requireActivity())
            .load(
                requireActivity().resources.getIdentifier(
                    curProduct.photo[curPic],
                    "drawable",
                    requireActivity().packageName
                )
            )
            .diskCacheStrategy(DiskCacheStrategy.NONE)
            .into(binding.ivPics)
    }
}